starlette/applications.py

```diff
@@ -109,7 +111,9 @@ def url_path_for(self, name: str, **path_params: str) -> URLPath:
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         scope["app"] = self
-        await self.middleware_stack(scope, receive, send)
+        task_group = scope["task_group"] = anyio.create_task_group()
```

Comment:
Would be interesting to know what quart chooses to do wrt. a request/response-wide task group. Eg. if you're running under trio do they provide an anyio TaskGroup or a trio Nursery?

Reply:
I've explored this elsewhere, and for an application lifespan to run background tasks with structured concurrency, it'd be better to have an app-wide task group from which all other tasks are spawned:
```python
diff --git a/starlette/applications.py b/starlette/applications.py
index 10bd075..168b78c 100644
--- a/starlette/applications.py
+++ b/starlette/applications.py
@@ -1,6 +1,7 @@
 import typing
 
 import anyio
+from anyio.abc import TaskGroup
 
 from starlette.datastructures import State, URLPath
 from starlette.exceptions import ExceptionMiddleware
@@ -37,6 +38,7 @@ class Starlette:
     Shutdown handler callables do not take any arguments, and may be be either
     standard functions, or async functions.
     """
+    task_group: TaskGroup
 
     def __init__(
         self,
@@ -111,8 +113,8 @@ class Starlette:
 
     async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
         scope["app"] = self
-        task_group = scope["task_group"] = anyio.create_task_group()
-        async with task_group:
+        self.task_group = anyio.create_task_group()
+        async with self.task_group:
             await self.middleware_stack(scope, receive, send)
 
     # The following usages are now discouraged in favour of configuration
diff --git a/starlette/middleware/base.py b/starlette/middleware/base.py
index 16d5064..4a7e58f 100644
--- a/starlette/middleware/base.py
+++ b/starlette/middleware/base.py
@@ -29,7 +29,7 @@ class BaseHTTPMiddleware:
     async def call_next(self, request: Request) -> Response:
         send_stream, recv_stream = anyio.create_memory_object_stream()
         scope = request.scope
-        task_group = scope["task_group"]
+        task_group = scope["app"].task_group
 
         async def coro() -> None:
             async with send_stream:
```

This then allows you to write a lifespan generator or startup/shutdown events such as:
```python
async def lifespan(app: Starlette):
    app.task_group.start_soon(background_task)
    ...
```

The only issue is, what happens if a task in that task group raises an error?

Reply:
> The only issue is, what happens if a task in that task group raises an error?

The lifespan task is cancelled and the server is shutdown. For the apps I write this is the desired behaviour.

Other users of Starlette might want a task wrapper that catches Exceptions and logs them instead

Reply:
here's a little demo of what quart_trio does:


```python
import httpx
import anyio
import quart
import quart_trio

import logging

logger = logging.getLogger(__name__)

blueprint = quart.Blueprint("blueprint", __name__)


class MyCrashError(Exception):
    pass


@blueprint.route("/")
async def index():
    async def crash():
        raise MyCrashError

    quart.current_app.nursery.start_soon(crash)
    return b""


def create_app():
    app = quart_trio.QuartTrio(__name__)
    app.register_blueprint(blueprint)
    return app


async def amain():
    async with anyio.create_task_group() as tg:
        tg.start_soon(create_app().run_task)
        async with httpx.AsyncClient() as client:
            await client.get("http://localhost:5000/")
            (await client.get("http://localhost:5000/")).raise_for_status()


anyio.run(amain, backend="trio")
```

```
[2021-05-13 19:04:22,398] Running on http://127.0.0.1:5000 (CTRL + C to quit)
[2021-05-13 19:04:22,414] 127.0.0.1:56808 GET / 1.1 200 0 2337
[2021-05-13 19:04:22,416] ERROR in app: Exception on request GET /
Traceback (most recent call last):
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart_trio/app.py", line 131, in handle_request
    return await self.full_dispatch_request(request_context)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart_trio/app.py", line 153, in full_dispatch_request
    result = await self.handle_user_exception(error)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart_trio/app.py", line 168, in handle_user_exception
    return await super().handle_user_exception(error)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart/app.py", line 943, in handle_user_exception
    raise error
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart_trio/app.py", line 151, in full_dispatch_request
    result = await self.dispatch_request(request_context)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart/app.py", line 1511, in dispatch_request
    return await self.ensure_async(handler)(**request_.view_args)
  File "/home/graingert/projects/demo-quart-trio/foo.py", line 22, in index
    quart.current_app.nursery.start_soon(crash)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/trio/_core/_run.py", line 988, in start_soon
    GLOBAL_RUN_CONTEXT.runner.spawn_impl(async_fn, args, self, name)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/trio/_core/_run.py", line 1429, in spawn_impl
    raise RuntimeError("Nursery is closed to new arrivals")
RuntimeError: Nursery is closed to new arrivals
[2021-05-13 19:04:22,414] ASGI Framework Lifespan error, continuing without Lifespan support
Traceback (most recent call last):
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/hypercorn/trio/lifespan.py", line 29, in handle_lifespan
    await invoke_asgi(self.app, scope, self.asgi_receive, self.asgi_send)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/hypercorn/utils.py", line 239, in invoke_asgi
    await app(scope, receive, send)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart/app.py", line 1697, in __call__
    await self.asgi_app(scope, receive, send)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart/app.py", line 1723, in asgi_app
    await asgi_handler(receive, send)
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/quart_trio/asgi.py", line 148, in __call__
    break
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/trio/_core/_run.py", line 815, in __aexit__
    raise combined_error_from_nursery
  File "/home/graingert/projects/demo-quart-trio/foo.py", line 20, in crash
    raise MyCrashError
MyCrashError
[2021-05-13 19:04:22,419] 127.0.0.1:56808 GET / 1.1 500 290 2909
Traceback (most recent call last):
  File "/home/graingert/projects/demo-quart-trio/foo.py", line 40, in <module>
    anyio.run(amain, backend="trio")
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/anyio/_core/_eventloop.py", line 55, in run
    return asynclib.run(func, *args, **backend_options)  # type: ignore
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/trio/_core/_run.py", line 1932, in run
    raise runner.main_task_outcome.error
  File "/home/graingert/projects/demo-quart-trio/foo.py", line 37, in amain
    (await client.get("http://localhost:5000/")).raise_for_status()
  File "/home/graingert/.virtualenvs/testing39/lib/python3.9/site-packages/httpx/_models.py", line 1405, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: 500 Server Error:  for url: http://localhost:5000/
For more information check: https://httpstatuses.com/500
```

Reply:
@graingert that's interesting.  A similar test app in Starlette would return "/" without error but raise an exception on the server, using uvicorn.

```python
from starlette.applications import Starlette
from starlette.responses import JSONResponse
from starlette.routing import Route

async def crash():
    raise Exception()

async def home(request):
    request.app.task_group.start_soon(crash)
    return JSONResponse({"ok": "True"})

app = Starlette(routes=[Route("/", home)])
```

```
INFO:     127.0.0.1:38388 - "GET / HTTP/1.1" 200 OK
ERROR:    Exception in ASGI application
Traceback (most recent call last):
  File "/home/jordan/src/starlette/venv/lib/python3.7/site-packages/uvicorn/protocols/http/h11_impl.py", line 396, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/jordan/src/starlette/venv/lib/python3.7/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "./starlette/applications.py", line 120, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/jordan/src/starlette/venv/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 528, in __aexit__
    raise exceptions[0]
  File "/home/jordan/src/starlette/venv/lib/python3.7/site-packages/anyio/_backends/_asyncio.py", line 559, in _run_wrapped_task
    await coro
  File "./testapp.py", line 6, in crash
    raise Exception()
Exception
```

Using hypercorn, it crashes and never responds.

Reply:
@uSpike I think this task_group should be scoped to the dispatch, and not attached to the app: https://github.com/uSpike/starlette/pull/1

Reply:
@graingert yep that makes sense to me, I'll get that in today.